core/src/main/c/share/txn_board.cpp

```diff
@@ -25,127 +25,127 @@
-    // 0 is initial value when shared memory is created
-    int64_t min = L_MAX;
-    T counts[];
-
-    inline static T inc(T val) {
-        return val + 1;
+    std::atomic<int64_t> max{0};
+    std::atomic<int64_t> min{0};
+    std::atomic<T> counts[];
+
+    template<typename TT>
+    inline static TT set_max_atomic(std::atomic<TT> &slot, TT value) {
+        TT current = slot.load();
+        while (value > current && !slot.compare_exchange_weak(current, value));
+        return std::max(value, current);
     }
 
-    inline static T dec(T val) {
-        return val - 1;
+    inline std::atomic<T> &get_counter(const int64_t offset) {
+        return counts[offset & mask];
     }
 
-public:
+    inline bool increment_count(int64_t txn) {
+        // Increment txn count
+        // but do not allow to use txn below max value
+        // Once there is count for txn 100
+        // there cannot be increments for txn 99, 98 etc
+        // When an attempt to acquire below max happens
+        // the attempt is rejected and TableReader should re-read _txn file
+        auto current_max = max.load();
+        if (txn < current_max) {
+            return false;
+        }
+        get_counter(txn)++; // atomic
 
-    inline int64_t get_min() {
-        return __atomic_load_n(&min, __ATOMIC_RELAXED);
-    }
+        while (!max.compare_exchange_weak(current_max, txn) && txn > current_max);
 
-    inline int64_t get_max() {
-        return __atomic_load_n(&max, __ATOMIC_RELAXED);
+        if (txn < current_max) {
+            // We cannot increment below max, only max or higher
+            // Roll back the increment
+            get_counter(txn)--; //atomic
+            return false;
+        }
+        return true;
     }
 
-    inline T *get_count_ptr(int64_t offset) {
-        return &(counts[offset & mask]);
+    inline int64_t update_min(const int64_t offset) {
+        int64_t o = min.load();
+        const int64_t check_limit = std::min(offset, o + size); // If no lock for full size, means not locks at all
+        while (o < check_limit && get_count(o) == 0) {
+            o++;
+        }
+        return set_max_atomic(min, o == check_limit ? offset : o);
     }
 
-    inline T get_count(int64_t offset) {
-        return __atomic_load_n(get_count_ptr(offset), __ATOMIC_RELAXED);
+public:
+
+    inline int64_t get_clean_min() {
+        int64_t val = min;
+        return val == L_MIN ? 0 : val;
     }
 
-    inline void update_min(const int64_t offset) {
-        int64_t o = get_min();
-        while (o < offset && get_count(o) == 0) {
-            o++;
-        }
-        set_max_atomic(&min, o);
+    inline T get_count(const int64_t &offset) {
+        return get_counter(offset);
     }
 
-    inline void txn_release(int64_t txn) {
-        const int64_t max_offset = get_max();
-        if (atomic_next(get_count_ptr(txn), dec) == 0 && get_min() == txn) {
-            update_min(max_offset);
+    inline T txn_release(int64_t txn) {
```

Comment:
Might be a good idea to make use of the implicit assertion for txn >= min as shown here:
https://github.com/puzpuzpuz/questdb/commit/6ab0a22b02d39b44322805d75b065cc7f88663ba#diff-14b580935a5db4b804f63bfd67cf563e563c1c5e6e9f5bd6e3faf030f4ae35b8R97-R107

This should help us with finding potential races in future.

Reply:
ok, done